mon-pix/app/services/locale.js

```diff
@@ -82,50 +103,121 @@ export default class LocaleService extends Service {
   }
 
   setCurrentLocale(locale) {
-    this.intl.setLocale(locale);
-    this.dayjs.setLocale(locale);
+    let nearestLocale = locale;
+
+    const localeCookieEnabled = this.featureToggles.featureToggles?.useLocale;
+    if (localeCookieEnabled) {
+      nearestLocale = this.#getNearestSupportedLocale(locale);
+      this.#setCookieLocale(nearestLocale);
+      this.__currentLocale = nearestLocale;
+    }
+
+    const language = this.#getLanguageFromLocale(nearestLocale);
+    this.intl.setLocale(language);
+    this.dayjs.setLocale(language);
 
     // metricsService may not be available for the different front apps
     const metricsService = getOwner(this).lookup('service:metrics');
     if (metricsService) {
-      metricsService.context.locale = locale;
+      metricsService.context.locale = nearestLocale;
+    }
+  }
+
+  detectBestLocale({ user, queryParams }) {
+    const localeCookieEnabled = this.featureToggles.featureToggles?.useLocale;
+
+    let locale;
+    if (localeCookieEnabled) {
+      locale = this.#detectBestLocale({ queryParams });
+    } else {
+      locale = this.#detectBestLocaleLegacy({ user, queryParams });
     }
+    this.setCurrentLocale(locale);
   }
 
-  detectBestLocale({ language, user }) {
+  #detectBestLocaleLegacy({ user, queryParams }) {
     if (this.currentDomain.isFranceDomain) {
-      this.setCurrentLocale(FRENCH_INTERNATIONAL_LOCALE);
-      this.#setLocaleCookie(FRENCH_FRANCE_LOCALE);
-      return;
+      this.#setCookieLocale(FRENCH_FRANCE_LOCALE);
+      return FRENCH_INTERNATIONAL_LOCALE;
     }
 
-    if (language) {
-      const supportedLanguage = this.#findSupportedLanguage(language);
-      this.setCurrentLocale(supportedLanguage);
-      return;
+    if (queryParams?.lang) {
+      return this.#getNearestSupportedLanguage(queryParams?.lang);
     }
 
     if (user?.lang) {
-      const supportedLanguage = this.#findSupportedLanguage(user.lang);
-      this.setCurrentLocale(supportedLanguage);
-      return;
+      return this.#getNearestSupportedLanguage(user.lang);
     }
 
-    this.setCurrentLocale(DEFAULT_LOCALE);
+    return DEFAULT_LOCALE;
   }
 
-  #findSupportedLanguage(language) {
-    return this.pixLanguages.filter((pixLanguage) => this.supportedLocales.includes(pixLanguage)).includes(language)
-      ? language
-      : DEFAULT_LOCALE;
+  #detectBestLocale({ queryParams }) {
+    const { isFranceDomain } = this.currentDomain;
+    const languageDetector = new LanguageDetector();
+
+    languageDetector.addDetector({
+      name: 'pix-domains',
+      lookup() {
+        if (isFranceDomain) return FRENCH_FRANCE_LOCALE;
```

Comment:
Suggestion pour ne pas définir une variable inutilement d’autant que cette variable n’est pas réutilisée dans la fonction : 

```suggestion
    const languageDetector = new LanguageDetector();

    languageDetector.addDetector({
      name: 'pix-domains',
      lookup() {
        if (this.currentDomain.isFranceDomain) return FRENCH_FRANCE_LOCALE;
```

Reply:
J'ai fait ça car le `this` n'est plus dans le même scope si on le mets ici et donc ne fonctionne pas.

Reply:
Ah oui ! Pardon !